File: v1/src/main/java/com/api/v1/utils/CustomerFinderUtil.java
package com.api.v1.utils;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.api.v1.domain.entities.Customer;
import com.api.v1.domain.repositories.CustomerRepository;
import com.api.v1.exceptions.customer.CustomerNotFoundError;

import reactor.core.publisher.Mono;

@Component
public class CustomerFinderUtil {

    @Autowired
    private CustomerRepository repository;

    public Mono<Customer> find(String ssn) {
        return repository
            .findAll()
            .filter(e -> e.getSsn().equals(ssn)
                && e.getArchivedAt() == null
            )
            .singleOrEmpty()
            .switchIfEmpty(Mono.error(new CustomerNotFoundError(ssn)));
    }
    
}

File: kotlin/v2/src/main/kotlin/com/api/v2/customer/utils/CustomerFInderUtil.kt
package com.api.v2.customer.utils

import com.api.v2.customer.domain.Customer
import com.api.v2.customer.domain.CustomerRepository
import com.api.v2.customer.exceptions.CustomerNotFoundException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.singleOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired

class CustomerFInderUtil {

    @Autowired
    private lateinit var customerRepository: CustomerRepository

    suspend fun find(ssn: String): Customer {
        return withContext(Dispatchers.IO) {
            val existingCustomer = customerRepository
                .findAll()
                .filter { e -> e.ssn == ssn }
                .singleOrNull()
            if (existingCustomer == null) {
                    throw CustomerNotFoundException()
                }
            existingCustomer
        }
    }

}